orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/v2/transform/V2ConfigStageInjectionTransform.java

```diff
@@ -0,0 +1,300 @@
+          templateStages.set(i, confStage);
+          log.debug("Template '{}' stage '{}' replaced by config {}", pipelineTemplate.getId(), confStage.getRefId(), templateConfiguration.getRuntimeId());
+        }
+      }
+    }
+  }
+
+  private void injectStages(V2PipelineTemplate pipelineTemplate) {
+    // Create initial graph via dependsOn. We need to include stages that the configuration defines with dependsOn as well
+    List<V2StageDefinition> initialStages = pipelineTemplate.getStages();
+
+    initialStages.addAll(templateConfiguration.getStages().stream()
+      .filter(s -> !s.getRequisiteStageRefIds().isEmpty())
+      .collect(Collectors.toList()));
+
+    pipelineTemplate.setStages(
+      createGraph(initialStages)
+    );
+
+    // Handle stage injections.
+    List<V2StageDefinition> templateInjectStages = injectStages(pipelineTemplate.getStages(), pipelineTemplate.getStages());
+    List<V2StageDefinition> configInjectStages = injectStages(templateConfiguration.getStages(), templateInjectStages);
+    pipelineTemplate.setStages(configInjectStages);
+  }
+
+  private enum Status {
+    VISITING,
+    VISITED
+  }
+
+  private static void dfs(String stageId,
+                          Set<V2StageDefinition> result,
+                          Map<String, Status> state,
+                          Map<String, V2StageDefinition> graph,
+                          Map<String, Integer> outOrder) {
+
+    state.put(stageId, Status.VISITING);
+    V2StageDefinition stage = graph.get(stageId);
+
+    if(stage == null) {
+      state.put(stageId, Status.VISITED);
+      return;
+    }
+
+    for (String n : stage.getRequisiteStageRefIds()) {
+      Status status = state.get(n);
+      if (status == Status.VISITING) {
+        throw new IllegalTemplateConfigurationException(
+          String.format("Cycle detected in graph (discovered on stage: %s)", stageId)
+        );
+      }
+
+      if (status == Status.VISITED) {
+        continue;
+      }
+
+      dfs(n, result, state, graph, outOrder);
+    }
+
+    result.add(stage);
+    state.put(stage.getRefId(), Status.VISITED);
+    stage.getRequisiteStageRefIds().forEach( s -> {
+      int order = outOrder.getOrDefault(s, 0);
+      outOrder.put(s, order + 1);
+    });
+  }
+
+  private static List<V2StageDefinition> createGraph(List<V2StageDefinition> stages) {
+    Set<V2StageDefinition> sorted = new LinkedHashSet<>();
+    Map<String, Status> state = new HashMap<>();
+
+//    stages.forEach(s -> s.setRequisiteStageRefIds(s.getDependsOn()));
+
+    Map<String, V2StageDefinition> graph = stages
+      .stream()
+      .collect(
+        Collectors.toMap(V2StageDefinition::getRefId, i -> i)
+      );
+
+    graph.forEach((k, v) -> dfs(k, sorted, state, graph, new HashMap<>()));
+
+    return new ArrayList<>(sorted);
+  }
+
+  private static List<V2StageDefinition> injectStages(List<V2StageDefinition> stages, List<V2StageDefinition> templateStages) {
```

Comment:
Given that we just rely on `refId` to order stages now, is this still needed?

Reply:
We also support the inject rules as well.